src/adapter/src/peek_client.rs

```diff
@@ -0,0 +1,381 @@
+    ) -> Result<&mut mz_compute_client::controller::instance::Client<Timestamp>, AdapterError> {
+        if !self.compute_instances.contains_key(&compute_instance) {
+            let client = self
+                .call_coordinator(|tx| Command::GetComputeInstanceClient {
+                    instance_id: compute_instance,
+                    tx,
+                })
+                .await?;
+            self.compute_instances.insert(compute_instance, client);
+        }
+        Ok(self
+            .compute_instances
+            .get_mut(&compute_instance)
+            .expect("ensured above"))
+    }
+
+    pub async fn ensure_oracle(
+        &mut self,
+        timeline: Timeline,
+    ) -> Result<&mut Arc<dyn TimestampOracle<Timestamp> + Send + Sync>, AdapterError> {
+        if !self.oracles.contains_key(&timeline) {
+            let oracle = self
+                .call_coordinator(|tx| Command::GetOracle {
+                    timeline: timeline.clone(),
+                    tx,
+                })
+                .await?;
+            self.oracles.insert(timeline.clone(), oracle);
+        }
+        Ok(self.oracles.get_mut(&timeline).expect("ensured above"))
+    }
+
+    /// Fetch a snapshot of the catalog for use in frontend peek sequencing.
+    /// Records the time taken in the adapter metrics, labeled by `context`.
+    pub async fn catalog_snapshot(&self, context: &str) -> Arc<Catalog> {
+        let start = std::time::Instant::now();
+        let CatalogSnapshot { catalog } = self
+            .call_coordinator(|tx| Command::CatalogSnapshot { tx })
+            .await;
+        self.coordinator_client
+            .metrics()
+            .catalog_snapshot_seconds
+            .with_label_values(&[context])
+            .observe(start.elapsed().as_secs_f64());
+        catalog
+    }
+
+    async fn call_coordinator<T, F>(&self, f: F) -> T
+    where
+        F: FnOnce(oneshot::Sender<T>) -> Command,
+    {
+        let (tx, rx) = oneshot::channel();
+        self.coordinator_client.send(f(tx));
+        rx.await.expect("sender dropped")
+    }
+
+    /// Acquire read holds on the given compute/storage collections, and
+    /// determine the smallest common valid write frontier among the specified collections.
+    ///
+    /// Similar to `Coordinator::acquire_read_holds` and `TimestampProvider::least_valid_write`
+    /// combined.
+    ///
+    /// Note: Unlike the Coordinator/StorageController's `least_valid_write` that treats sinks
+    /// specially when fetching storage frontiers (see `mz_storage_controller::collections_frontiers`),
+    /// we intentionally do not special‑case sinks here because peeks never read from sinks.
+    /// Therefore, using `StorageCollections::collections_frontiers` is sufficient.
+    ///
+    /// Note: self is taken &mut because of the lazy fetching in `get_compute_instance_client`.
+    pub async fn acquire_read_holds_and_collection_write_frontiers(
```

Comment:
The name is a bit misleading since the method doesn't return the collection write frontiers but just the minimum of them.

Reply:
Indeed, changed to `acquire_read_holds_and_least_valid_write`.